sqltoken/tokenize.go

```diff
@@ -0,0 +1,730 @@
+	return Tokenize(s, MySQLConfig())
+}
+
+func TokenizePostgreSQL(s string) Tokens {
+	return Tokenize(s, PostgreSQLConfig())
+}
+
+const debug = false
+
+// Tokenize breaks up SQL strings into Token objects.  No attempt is made
+// to break successive punctuation.
+func Tokenize(s string, config Config) Tokens {
+	if len(s) == 0 {
+		return []Token{}
+	}
+	tokens := make([]Token, 0, len(s)/5)
+	tokenStart := 0
+	var i int
+	var firstDollarEnd int
+
+	// Why is this written with Goto you might ask?  It's written
+	// with goto because RE2 can't handle complex regex and PCRE
+	// has external dependencies and thus isn't friendly for libraries.
+	// So, it could have had a switch with a state variable, but that's
+	// just a way to do goto that's lower performance.  Might as
+	// well do goto the natural way.
+
+	token := func(t TokenType) {
+		if debug {
+			fmt.Printf("> %s: {%s}\n", t, s[tokenStart:i])
+		}
+		if i-tokenStart == 0 {
+			return
+		}
+		if len(tokens) > 0 && tokens[len(tokens)-1].Type == t && combineOkay(t) {
+			tokens[len(tokens)-1].Text += s[tokenStart:i]
+		} else {
+			tokens = append(tokens, Token{
+				Type: t,
+				Text: s[tokenStart:i],
+			})
+		}
+		tokenStart = i
+	}
+
+BaseState:
+	for i < len(s) {
+		c := s[i]
+		i++
+		switch c {
+		case '/':
+			if i < len(s) && s[i] == '*' {
+				goto CStyleComment
+			}
+			token(Punctuation)
+		case '\'':
+			goto SingleQuoteString
+		case '"':
+			goto DoubleQuoteString
+		case '-':
+			if i < len(s) && s[i] == '-' {
+				goto SkipToEOL
+			}
+			token(Punctuation)
+		case '#':
+			if config.NoticeHashComment {
+				goto SkipToEOL
+			}
+			token(Punctuation)
+		case ';':
+			token(Semicolon)
+		case '?':
+			if config.NoticeQuestionMark {
+				token(QuestionMark)
+			} else {
+				token(Punctuation)
+			}
+		case ' ', '\n', '\r', '\t', '\b', '\v', '\f':
+			goto Whitespace
+		case '.':
+			goto PossibleNumber
+		case '~', '`', '!', '%', '^', '&', '*', '(', ')', '+', '=', '{', '}', '[', ']',
+			'|', '\\', ':', '<', '>', ',':
+			token(Punctuation)
+		case '$':
+			// $1
+			// $seq$ stuff $seq$
+			// $$stuff$$
+			if config.NoticeDollarQuotes || config.NoticeDollarNumber {
+				goto Dollar
+			}
+			token(Punctuation)
+		case 'U':
+			// U&'d\0061t\+000061'
+			if config.NoticeUAmpPrefix && i+1 < len(s) && s[i] == '&' && s[i+1] == '\'' {
+				i += 2
+				goto SingleQuoteString
+			}
+			goto Word
+		case 'x', 'X':
+			// X'1f' x'1f'
+			if config.NoticeHexNumbers && i < len(s) && s[i] == '\'' {
+				i++
+				goto QuotedHexNumber
+			}
+			goto Word
+		case 'b', 'B':
+			if config.NoticeBinaryNumbers && i < len(s) && s[i] == '\'' {
+				i++
+				goto QuotedBinaryNumber
+			}
+			goto Word
+		case 'a' /*b*/, 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm',
+			'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w' /*x*/, 'y', 'z',
+			'A' /*B*/, 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M',
+			'N', 'O', 'P', 'Q', 'R', 'S', 'T' /*U*/, 'V', 'W' /*X*/, 'Y', 'Z',
+			'_':
```

Comment:
This seems confusing and hard to maintain. At least there should be a comment explaining these values. It may make sense to handle these as ranges outside the `switch`.

Reply:
This is a performance hack: anything missed will be caught by the unicode code path below.  See added comment.